### ISSUE TITLE
`VectorTools::Interpolate()` gives different results for `FE_DGQ` and `FE_RaviartThomasNodal`

### ISSUE THREAD

**Issue by NiklasWik** [dealii/dealii#13580]
Using the following function:
```C++
  double
  value(dealii::Point<dim> const & p, unsigned int const component = 0) const
  {
    double result = 0.0;

    if(component == 0)
      result = std::sin(p[0]) * std::cos(p[1]) * std::cos(p[2]);
    else if(component == 1)
      result = std::cos(p[0]) * std::sin(p[1]) * std::cos(p[2]);
    else if(component == 2)
      result = std::cos(p[0]) * std::cos(p[1]) * std::sin(p[2]);

    return result;
  }
```
while calling `VectorTools::Interpolate()` for `FE_DGQ` and `FE_RaviartThomasNodal` leads to following results

DGQ: 
 x-, y- and z-component
<img src="https://user-images.githubusercontent.com/62885387/160376439-cc1ff9d9-ac98-46e6-b781-42ad5a234998.png" width="200">  <img src="https://user-images.githubusercontent.com/62885387/160376441-6deaf707-a706-400e-80f1-5b18b62cc3c7.png" width="200">  <img src="https://user-images.githubusercontent.com/62885387/160376446-e126234a-8f7d-4995-ba0a-f95200fa5fac.png" width="200">

RT: 
x-, y-, and z-component
<img src="https://user-images.githubusercontent.com/62885387/160376748-77b6d8d9-243d-4638-9bda-4be2cb81a145.png" width=200>  <img src="https://user-images.githubusercontent.com/62885387/160376750-daf73dc4-99f5-4183-a491-7cf1e839ff8b.png" width="200">  <img src="https://user-images.githubusercontent.com/62885387/160376753-ce44504d-5d72-4737-badb-7bce1491f2ea.png" width="200">

The interpolation gives the same result for the x-component, but flips the coordinates for the y- and z-component for the RT element.

**Comment by peterrum** [comment]
How does the result of `VectorTools::project()` looks like? I am sure if is possible to interpolate in the case of `FE_RaviartThomasNodal` (how do the support points looks like?)

**Comment by kronbichler** [comment]
Indeed, it is likely the location of the support points that is wrong and that we might want to fix. But in general the `interpolate` function will not work properly because even the nodal version of the RT element is not interpolatory: On the one hand, different components have different support points (which might still be ok in the simple case you are showing, but already leads to trouble for more complicated functions), but more importantly, the Piola transform that is used in the RT case will be missing in the case of interpolation. In your pictures, you might be able to see this: If you change the mesh size, you will see that the magnitude of the solution changes.
So one really needs a projection here, as interpolation won't work. The downside is that `VectorTools::project` is currently not prepared for that case in an optimal way, as we do not take the parallel and fast matrix-free path that gets used for simple `FE_Q` / `FE_DGQ` elements. Otherwise, the code is not parallelized as it uses a sparse matrix to set up the projection and pretty slow.

**Comment by nfehn** [comment]
> I am sure if is possible to interpolate in the case of `FE_RaviartThomasNodal`

The code does something for `interpolate()` (it does not throw an exception), so why should interpolate not be possible?

**Comment by nfehn** [comment]
But do we agree that the code at least has to assert this, with an error in both debug and release mode?

**Comment by nfehn** [comment]
> which might still be ok in the simple case you are showing, but already leads to trouble for more complicated functions

What does "simple"/"complicated" mean in this context?

**Comment by nfehn** [comment]
> Indeed, it is likely the location of the support points that is wrong and that we might want to fix

What do you mean with "might want to fix"? Does the problem affect other parts of the dealii code as well? Reading the current discussion, it appears to be most appropriate to `AsserThrow(false, ...)` the RT-case in `VectorTools::interpolate()`.

**Comment by kronbichler** [comment]
It helps to look at the implementation of `VectorTools::interpolate`: I realize that this might actually still work for Hdiv and Hcurl functions with generalized support points (as long as we can evaluate the functions properly), as the code aims to compensate for that transformation: https://github.com/dealii/dealii/blob/b0ece6447f96a65f33a269551c61d1dc02e61e76/include/deal.II/numerics/vector_tools_interpolate.templates.h#L344-L385
I did not look more closely into the code and whether it really works, but at least I am somewhat optimistic that it could be used after some modification. The essential part is to (a) check what https://github.com/dealii/dealii/blob/b0ece6447f96a65f33a269551c61d1dc02e61e76/include/deal.II/numerics/vector_tools_interpolate.templates.h#L384-L385 does, i.e., follow down that function and see what it does, and (b) to make sure that the points as set up by https://github.com/dealii/dealii/blob/b0ece6447f96a65f33a269551c61d1dc02e61e76/source/fe/fe_raviart_thomas_nodal.cc#L384-L401 are correct in terms of how we understand the polynomials.

**Comment by kronbichler** [comment]
That said, `FE_RaviartThomasNodal` has not been widely used inside deal.II yet, so it is not surprising that new use cases uncover small mistakes or conceptual problems here and there.

**Comment by kronbichler** [comment]
> What does "simple"/"complicated" mean in this context?

I was thinking about a function where we can only evaluate all `dim` vector components in a point at once, as opposed to evaluating the component `d` in a certain point and other components in other points. As said in the other message, that might not actually be the distinction I was suggesting because I did not read carefully what the current implementation of `interpolate` does. From my somewhat more careful look, it seems it always takes all `dim` components of the function in all points of `dim` components of the basis functions, and it is in the `transform` or `convert_generalized_...` functions that things get filtered out appropriately. That would need to be checked carefully with the given code. In particular since the figures suggest different scaling in the obtained solution of the DG and RT cases, respectively.

**Comment by bangerth** [comment]
Just to confirm, at least in principle using generalized support points should be able to deal with this situation. In other words, I *think* that the function should work; if it doesn't, that's a bug.

**Comment by NiklasWik** [comment]
I made this change 880be34008c3a73dd4632f9867a2b2995adccc39 to the renumbering of the anisotropic points in the case of `dim=3`. The plot I get after using `VectorTools::project()` is still the same as before, but now it also corresponds to the result from solving the mass matrix system using the matrixfree implementation I've been working on. Before, the latter solution was flipped in the x and z coordinates for the second component after plotting, in the same way I felt that the `project()` flipped the solution before plotting.

I'm unsure if this change is correct since I don't have enough knowledge about how the support points work. From what I can tell, somewhere the x and z coordinates flip (I've also encountered this while implementing the face evaluation kernals for the matrixfree implementation I mentioned). However, since the points are symmetric in x and z direction for the second component, makes this seem ok... I'm also unsure if this brakes anything else as I don't know what tests to run.

**Comment by kronbichler** [comment]
@NiklasWik I guess this was resolved by #13586, right?

**Comment by NiklasWik** [comment]
> @NiklasWik I guess this was resolved by #13586, right?

Pretty sure it wasn't. I can double check.

**Comment by NiklasWik** [comment]
The y-component now looks like this for the RT case, given the same sin-function as above. 
<img src="https://user-images.githubusercontent.com/62885387/167575460-d5232963-915b-42c3-917b-aa759ec8a467.png" width="300">

Also, I tried to use `VectorTools::compute_nonzero_normal_flux_constraints` with RT as well and it gave me this exception 
<img src="https://user-images.githubusercontent.com/62885387/167576403-596f37cf-415c-4a64-8e4c-6880f965a459.png" width=500>
Not sure if related or if I should open another issue?

**Comment by peterrum** [comment]
> Edit: Realized that I should use VectorTools::project_boundary_values_div_conforming instead which seems to work for FE_RaviartThomasNodal as well, as apposed to FE_RaviartThomas which it is made for.

@NiklasWik What is working and what is not? `interpolate()`? `project()`? `project_boundary_values_div_conforming()`? It would be great to have `interpolate()` and one of the `project()` functions working. Maybe we could also extend the documentation of `VectorTools::project_boundary_values_div_conforming` so that it states that it also works for `FE_RaviartThomasNodal`? Adding some assert might be also an option.

**Comment by NiklasWik** [comment]
`interpolate()` does not work. 

`project()` works.

I have only tried `project_boundary_values_div_conforming()` with a zero-function, which works. Here I had to change the function as it previously asserts if `FE_RaviartThomas` is used. I can create the pr later.

**Comment by bangerth** [comment]
What should we do about this in view of the upcoming release? Can we fix it before the release, or does it have to be postponed?

**Comment by kronbichler** [comment]
I will have a look on this.

**Comment by tamiko** [comment]
@kronbichler @bangerth I need a call on this issue regarding the release. :slightly_smiling_face:

**Comment by tamiko** [comment]
I do not think that this fix will happen before the Thursday - I am reassigning to 9.5. We can simply cherry pick to the release branch next week.

**Comment by kronbichler** [comment]
I pushed a fix in #13976. It is a small adjustment and should be safe.

**Comment by kronbichler** [comment]
> I have only tried `project_boundary_values_div_conforming()` with a zero-function, which works. Here I had to change the function as it previously asserts if `FE_RaviartThomas` is used. I can create the pr later.

That would be great.

**Comment by NiklasWik** [comment]
> That would be great.

Already merged! #13907

**Comment by kronbichler** [comment]
But then me looking at the wrong place was not for vein, since I noticed this one:
https://github.com/dealii/dealii/blob/0ab370fc8a9a95efd51167a788df28a276d29d8b/include/deal.II/numerics/vector_tools_boundary.templates.h#L2652-L2654
and
https://github.com/dealii/dealii/blob/0ab370fc8a9a95efd51167a788df28a276d29d8b/include/deal.II/numerics/vector_tools_boundary.templates.h#L2600-L2602
further down in the file, compared to your changes. Let me propose an update.
